pkg/admission/server_test.go

```diff
@@ -120,6 +120,208 @@ func TestServeHTTPSubmissions(t *testing.T) {
+				reqBody: dedent.Dedent(`{
+						"kind": "AdmissionReview",
+						"apiVersion": "` + apiVersion + `",
+						"request": {
+							"uid": "7313cd05-eddc-4150-b88c-971a0d53b2ab",
+							"resource": {
+								"group": "networking.x-k8s.io",
+								"version": "v1alpha2",
+								"resource": "gateways"
+							},
+							"object": {
+   								"kind": "Gateway",
+   								"apiVersion": "networking.x-k8s.io/v1alpha1",
+   								"metadata": {
+   								   "name": "gateway-1",
+   								   "labels": {
+   								      "app": "foo"
+   								   }
+   								},
+   								"spec": {
+									"gatewayClassName": "contour-class",
+									"listeners": [
+										{
+											"port": 80,
+											"protocol": "HTTP",
+											"hostname": "foo.com",
+											"routes": {
+												"group": "networking.x-k8s.io",
+												"kind": "HTTPRoute",
+												"namespaces": {
+													"from": "All"
+												}
+											}
+										}
+									]
+   								}
+							},
+						"operation": "CREATE"
+						}
+					}`),
+				wantRespCode: http.StatusOK,
+				wantSuccessResponse: admission.AdmissionResponse{
+					UID:     "7313cd05-eddc-4150-b88c-971a0d53b2ab",
+					Allowed: true,
+					Result:  &metav1.Status{},
+				},
+			},
+			{
+				name: "invalid v1alpha2 Gateway resource with bad listener hostname",
+				reqBody: dedent.Dedent(`{
+						"kind": "AdmissionReview",
+						"apiVersion": "` + apiVersion + `",
+						"request": {
+							"uid": "7313cd05-eddc-4150-b88c-971a0d53b2ab",
+							"resource": {
+								"group": "networking.x-k8s.io",
+								"version": "v1alpha2",
+								"resource": "gateways"
+							},
+							"object": {
+   								"kind": "Gateway",
+   								"apiVersion": "networking.x-k8s.io/v1alpha1",
+   								"metadata": {
+   								   "name": "gateway-1",
+   								   "labels": {
+   								      "app": "foo"
+   								   }
+   								},
+   								"spec": {
+									"gatewayClassName": "contour-class",
+									"listeners": [
+										{
+											"port": 80,
+											"protocol": "HTTP",
+											"hostname": "!@!.foo.com",
+											"routes": {
+												"group": "networking.x-k8s.io",
+												"kind": "HTTPRoute",
+												"namespaces": {
+													"from": "All"
+												}
+											}
+										}
+									]
+   								}
+							},
+						"operation": "CREATE"
+						}
+					}`),
+				wantRespCode: http.StatusOK,
+				wantSuccessResponse: admission.AdmissionResponse{
+					UID:     "7313cd05-eddc-4150-b88c-971a0d53b2ab",
+					Allowed: false,
+					Result: &metav1.Status{
+						Code:    400,
+						Message: "spec.listeners[0].hostname: Invalid value: \"!@!.foo.com\": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')",
```

Comment:
Can/Should we DRY out `a lowercase RFC...`?
Maybe define a string at the test package level and user that here and one other place we have?

Reply:
Ok I pulled the hostname + error message out in a const. Is this what you were thinking @hbagdi?